OSDOCS-2457: document procedures for Global Options to Enable HTTP Strict Transport Security (HSTS) (#35812)
openshift/openshift-docs: modules/nw-enabling-hsts-per-route.adoc

// Module included in the following assemblies:
// * networking/configuring-routing.adoc

[id="nw-enabling-hsts-per-route_{context}"]
= Enforcing HTTP strict transport security per-route

HTTP strict transport security (HSTS) is implemented in the HAProxy template and applied to edge and reencrypt routes that have the `haproxy.router.openshift.io/hsts_header` annotation.

> If you do not have many routes or regulatory compliance requirements, you can use the per-route method of configuring HSTS.

Reviewer: ```suggestion
```

This is not accurate.  They have to use the per-route method no matter what.

Author: Thank you! I missed that one--will remove.

.Prerequisites

* You are logged in to the cluster with a user with `cluster-admin` privileges.

.Procedure

* To enforce HSTS on a route, add the `haproxy.router.openshift.io/hsts_header` value to the edge terminated or re-encrypt route::
+
.Example route configured with an annotation
+
[source,yaml]
----
apiVersion: v1
kind: Route
metadata:
  annotations:
    haproxy.router.openshift.io/hsts_header: max-age=31536000;includeSubDomains;preload <1> <2> <3>
...
spec:
  host: def.abc.com
  tls:
    termination: "reencrypt"
    ...
  wildcardPolicy: "Subdomain"
----

> <1> Required. `max-age` measures the length of time, in seconds, that the HSTS policy is in effect. The client updates `max-age` whenever a response with a HSTS header is received from the host. When `max-age` times out, the client discards the policy. If set to `0`, it negates the policy.

Reviewer: ```suggestion
<1> Required. `max-age` measures the length of time, in seconds, that the HSTS policy is in effect. If set to `0`, it negates the policy.
```
IMHO including the other information will be confusing.

<2> Optional. When included, `includeSubDomains` tells the client

> that all subdomains of the host are to be treated the same as the host.

Reviewer: ```suggestion
that all subdomains of the host should have the same HSTS policy as the host.
```

> <3> Optional. When `max-age` is greater than 0, you can add `preload` in  `haproxy.router.openshift.io/hsts_header` to allow external services to include this site in their HSTS preload lists. For example, sites such as Google can construct a list of sites that have `preload` set. Browsers can then use these lists to determine which sites they can communicate with over HTTPS, before they have interacted with the site. Without `preload` set, browsers must have interacted with the site over HTTPS to get the header.

Reviewer: ```suggestion
<3> Optional. When `max-age` is greater than 0, you can add `preload` in  `haproxy.router.openshift.io/hsts_header` to allow external services to include this site in their HSTS preload lists.  Browsers can then use these lists to determine which sites they can communicate with over HTTPS, even before they have interacted with the site. Without `preload` set, browsers must have interacted with the site over HTTPS, at least once, to get the header.
```
